pkg/motor/connect.go

```diff
@@ -0,0 +1,206 @@
+		return nil, err
+	}
+
+	// Utilize discovery protocol
+	if request.GetEnableDiscovery() {
+		mtr.log.Info("Enabling Discovery...")
+		mtr.discover, err = dp.New(context.Background(), mtr.SonrHost, mtr.callback)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Utilize transmit protocol
+	if request.GetEnableTransmit() {
+		mtr.log.Info("Enabling Transmit...")
+		mtr.transmit, err = tp.New(context.Background(), mtr.SonrHost)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	mtr.log.Info("✅ Motor Host Connected")
+	mtr.hostInitialized = true
+	return &mt.ConnectResponse{
+		Success: true,
+		Message: "Successfully connected host to network",
+	}, nil
+}
+
+func (mtr *motorNodeImpl) OpenLinking(request mt.LinkingRequest) (*mt.LinkingResponse, error) {
+	if !mtr.IsHostActive() {
+		return nil, fmt.Errorf("host inactive")
+	}
+
+	// Setup stream handler
+	mtr.SonrHost.Host().SetStreamHandler(proto.ProtoIDv1, mtr.handleLinking)
+
+	peerInfo := &peer.AddrInfo{
+		ID:    mtr.SonrHost.Host().ID(),
+		Addrs: mtr.SonrHost.Host().Addrs(),
+	}
+	addrs, err := peer.AddrInfoToP2pAddrs(peerInfo)
+	if err != nil {
+		return nil, fmt.Errorf("convert to p2p addrs: %s", err)
+	}
+
+	encodedAddrs := make([][]byte, len(addrs))
+	for i, a := range addrs {
+		encodedAddrs[i] = a.Bytes()
+	}
+
+	return &mt.LinkingResponse{
+		Success:  true,
+		P2PAddrs: encodedAddrs,
+	}, nil
+}
+
+func (m *motorNodeImpl) PairDevice(request mt.PairingRequest) (*mt.PairingResponse, error) {
+	if !m.IsHostActive() {
+		return nil, fmt.Errorf("host is not active")
+	}
+	if len(request.P2PAddrs) == 0 {
+		return nil, fmt.Errorf("missing p2p addresses")
+	}
+
+	var err error
+	addrs := make([]ma.Multiaddr, len(request.P2PAddrs))
+	for i, a := range request.P2PAddrs {
+		addrs[i], err = ma.NewMultiaddrBytes(a)
+		if err != nil {
+			return nil, fmt.Errorf("decode p2p multiaddr: %s", err)
+		}
+	}
+
+	peerInfos, err := peer.AddrInfosFromP2pAddrs(addrs...)
+	if err != nil {
+		return nil, fmt.Errorf("get p2p addrs: %s", err)
+	}
+
+	for _, peerInfo := range peerInfos {
+		err = m.connectToPeerAndTransmit(peerInfo, request)
+		if err == nil {
+			break
+		}
+	}
+	if err != nil {
+		return nil, fmt.Errorf("attempt to connect to all peerInfos: %s", err)
+	}
+
+	return &mt.PairingResponse{
+		Success: true,
+	}, nil
+}
+
+func (mtr *motorNodeImpl) connectToPeerAndTransmit(peerInfo peer.AddrInfo, request mt.PairingRequest) error {
+	if err := mtr.SonrHost.Connect(peerInfo); err != nil {
+		mtr.SonrHost.Host().Peerstore().ClearAddrs(peerInfo.ID)
+		return errors.Wrap(err, "failed to connect to peer while attempting to pair")
+	}
+
+	str, err := mtr.SonrHost.NewStream(context.Background(), peerInfo.ID, proto.ProtoIDv1)
+	if err != nil {
+		return errors.Wrap(err, "failed to open stream while attempting to pair")
+	}
+	defer str.Close()
+
+	strWr := msgio.NewWriter(str)
+	authInfo := &ct.AuthInfo{
+		Address:   mtr.Address,
+		Did:       mtr.GetDID().String(),
+		AesPskKey: mtr.encryptionKey,
```

Comment:
is this assuming the motor has already been logged in? we might want to check if the `encryptionKey` is nil

Reply:
Yes, you must be logged in in order to add a new device. Adding a check for this at the beginning of the function.